homeassistant/components/smtp/notify.py

```diff
@@ -223,14 +223,80 @@ def _send_email(self, msg, recipients):
             try:
                 mail.sendmail(self._sender, recipients, msg.as_string())
                 break
-            except smtplib.SMTPServerDisconnected:
+            except smtplib.SMTPServerDisconnected as err:
                 _LOGGER.warning(
-                    "SMTPServerDisconnected sending mail: retrying connection"
+                    "SMTPServerDisconnected while sending mail: Retrying Connection"
                 )
+                _LOGGER.debug("SMTPServerDisconnected Full Error Contents: %s", err)
                 mail.quit()
                 mail = self.connect()
-            except smtplib.SMTPException:
-                _LOGGER.warning("SMTPException sending mail: retrying connection")
+            except smtplib.SMTPSenderRefused as err:
+                _LOGGER.warning(
+                    "SMTPSenderRefused while sending mail: Retrying Connection --- "
+                    "Error Code: %s  - Error Message: %s",
+                    err.smtp_code,
+                    err.smtp_error,
+                )
+                _LOGGER.debug("SMTPSenderRefused Full Error Contents: %s", err)
+                mail.quit()
+                mail = self.connect()
+            except smtplib.SMTPRecipientsRefused as err:
+                _LOGGER.warning(
+                    "SMTPRecipientsRefused while sending mail: Retrying Connection"
+                )
+                _LOGGER.debug("SMTPRecipientsRefused Full Error Contents: %s", err)
+                mail.quit()
+                mail = self.connect()
+            except smtplib.SMTPDataError as err:
```

Comment:
A lot of these catch clauses are very similar. You could catch many of them in a single except block (or perhaps just catch `smtplib.SMTPResponseException` which all the exceptions with `smtp_code` and `smtp_error` derive from), and include the exception type name in the formatted strings (e.g. `type(err).__name__`) without losing too much logging information, and reducing the verbosity of the code a lot.

Reply:
You're right, I was able to drastically reduce the code by just catching them all at once and dynamically setting the messages based on the exceptions. I figured I'd still catch the four that have `smtplib.SMTPResponseException` as their base exception so that I could reference them specifically by name.